br/pkg/checksum/executor.go

```diff
@@ -0,0 +1,321 @@
+)
+
+type Checksum struct {
+	Crc64Xor   uint64
+	TotalKvs   uint64
+	TotalBytes uint64
+}
+
+func (c *Checksum) UpdateChecksum(crc64Xor, totalKvs, totalBytes uint64) {
+	c.Crc64Xor ^= crc64Xor
+	c.TotalKvs += totalKvs
+	c.TotalBytes += totalBytes
+}
+
+func (c *Checksum) String() string {
+	return fmt.Sprintf("crc64xor:%d, totalKvs:%d, totalBytes:%d",
+		c.Crc64Xor, c.TotalKvs, c.TotalBytes)
+}
+
+type StorageChecksumMethod int32
+
+const (
+	StorageChecksumCommand StorageChecksumMethod = 0
+	StorageScanCommand     StorageChecksumMethod = 1
+)
+
+// ExecutorBuilder is used to build
+type Executor struct {
+	startKey    []byte
+	endKey      []byte
+	apiVersion  kvrpcpb.APIVersion
+	pdClient    pd.Client
+	concurrency uint
+	mutex       sync.Mutex // lock for update checksum
+}
+
+// NewExecutorBuilder returns a new executor builder.
+func NewExecutor(startKey, endKey []byte, apiVersion kvrpcpb.APIVersion, pdClient pd.Client,
+	concurrency uint) *Executor {
+	return &Executor{
+		startKey:    startKey,
+		endKey:      endKey,
+		apiVersion:  apiVersion,
+		pdClient:    pdClient,
+		concurrency: concurrency,
+	}
+}
+
+func adjustRegionRange(startKey, endKey, regionStart, regionEnd []byte) ([]byte, []byte) {
+	retStart := startKey
+	if string(retStart) < string(regionStart) {
+		retStart = regionStart
+	}
+	retEnd := endKey
+	if len(retEnd) == 0 || (len(regionEnd) != 0 && string(retEnd) > string(regionEnd)) {
+		retEnd = regionEnd
+	}
+	return retStart, retEnd
+}
+
+func (exec *Executor) doChecksumOnRegion(
+	ctx context.Context,
+	regionInfo *restore.RegionInfo,
+	splitClient restore.SplitClient,
+	finalChecksum *Checksum,
+	progressCallBack func(backup.ProgressUnit),
+) error {
+	var peer *metapb.Peer
+	if regionInfo.Leader != nil {
+		peer = regionInfo.Leader
+	} else {
+		if len(regionInfo.Region.Peers) == 0 {
+			return errors.Annotate(berrors.ErrRestoreNoPeer, "region does not have peer")
+		}
+		peer = regionInfo.Region.Peers[0]
+	}
+	storeID := peer.GetStoreId()
+	store, err := splitClient.GetStore(ctx, storeID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	conn, err := grpc.Dial(store.GetAddress(), grpc.WithInsecure())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer conn.Close()
+
+	client := tikvpb.NewTikvClient(conn)
+	rangeStart, rangeEnd := adjustRegionRange(exec.startKey, exec.endKey, regionInfo.Region.StartKey, regionInfo.Region.EndKey)
+	apiver := exec.apiVersion
+	if apiver == kvrpcpb.APIVersion_V1TTL {
+		apiver = kvrpcpb.APIVersion_V1
+	}
+	resp, err := client.RawChecksum(ctx, &kvrpcpb.RawChecksumRequest{
+		Context: &kvrpcpb.Context{
+			RegionId:    regionInfo.Region.Id,
+			RegionEpoch: regionInfo.Region.RegionEpoch,
+			Peer:        peer,
+			ApiVersion:  apiver,
+		},
+		Algorithm: kvrpcpb.ChecksumAlgorithm_Crc64_Xor,
+		Ranges: []*kvrpcpb.KeyRange{{
+			StartKey: rangeStart,
+			EndKey:   rangeEnd,
+		}},
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if resp.GetRegionError() != nil {
```

Comment:
Some errors are inevitable, e.g. `not_leader`. Maybe we can refer to how does `client-go` deal with these errors..

Reply:
Change to use client-go to scan the data. `client-go` now does not support `checksum`, so add retry for region error here. In the future, i think it's better to support checksum in `client-go` and reuse the error handling code.

Reply:
Yes. Add checksum to client-go would be better.

Reply:
emm, if `BR` depend on the latest `client-go`, we need resolve the dependency problem between `grpc` and `etcd` in different versions. I have another branch https://github.com/haojinming/migration/tree/br-checksum-with-client to do this and need upgrade to go1.18 to do this as `BR` depend on `tidb` repos and `tidb` use go1.18 version now. Still have some problem in `go` static check. So I would like to delay this.